sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -148,6 +149,8 @@ private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
 
 object SharedState {
 
+  URL.setURLStreamHandlerFactory(new SparkUrlStreamHandlerFactory())
```

Comment:
We can simply call `URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
` and everything works happily. The only problem is, `URL.setURLStreamHandlerFactory` can only be called once per JVM (check javadoc of URL class), and if we want to support more stream handlers in the future, we wouldn't be able to call `URL.setURLStreamHandlerFactory` again. It's like you only have one wall plug but you have several laptops and you'll have to use a power strip.